hkmc2/shared/src/main/scala/hkmc2/codegen/StackSafeTransform.scala

```diff
@@ -121,40 +103,72 @@ class StackSafeTransform(depthLimit: Int, paths: HandlerPaths)(using State):
-      methods.map(rewriteFn),
-      privateFields,
-      publicFields, rewriteBlk(preCtor),
-      rewriteBlk(ctor),
-      mod.map(rewriteObjBody(_, isTopLevel)),
-    )
+  def rewriteCls(defn: ClsLikeDefn, isTopLevel: Bool): ClsLikeDefn = defn.parentPath match
+    case Some(value) if value eq paths.contClsPath => defn
+    case _ =>
+      val ClsLikeDefn(owner, isym, sym, k, paramsOpt, auxParams,
+        parentPath, methods, privateFields, publicFields, preCtor, ctor, mod) = defn
+      ClsLikeDefn(
+        owner, isym, sym, k, paramsOpt, auxParams, parentPath,
+        methods.map(rewriteFn),
+        privateFields,
+        publicFields, 
+        rewriteBlk(preCtor, L(BlockMemberSymbol("TODO", Nil)), 1), // TODO: preCtor is not translated in handler lowering
+        if isTopLevel && (defn.k is syntax.Mod) then transformTopLevel(ctor) else rewriteBlk(ctor, R(isym), 1),
+        mod.map(rewriteObjBody(_, isTopLevel)),
+      )
   
   def rewriteObjBody(defn: ClsLikeBody, isTopLevel: Bool): ClsLikeBody =
     ClsLikeBody(
       defn.isym,
       defn.methods.map(rewriteFn),
       defn.privateFields,
       defn.publicFields,
-      if isTopLevel then transformTopLevel(defn.ctor) else rewriteBlk(defn.ctor),
-    )
+      if isTopLevel then transformTopLevel(defn.ctor) else rewriteBlk(defn.ctor, R(defn.isym), 1),
+      )
 
-  def rewriteBlk(blk: Block) =
-    val curDepth =
+  // fnOrCls points us to the doUnwind function
+  def rewriteBlk(blk: Block, fnOrCls: FnOrCls, increment: Int) =
+    var usedDepth = false
+    lazy val curDepth =
+      usedDepth = true
       TempSymbol(None, "curDepth")
+      
+    val doUnwindPath = doUnwindMap.get(fnOrCls)
     val newBody = transform(blk, curDepth)
+    
     if isTrivial(blk) then
       newBody
-    else
+    else if doUnwindPath.isEmpty then
       val resSym = TempSymbol(None, "stackDelayRes")
       blockBuilder
-        .assign(curDepth, stackDepthPath)
+        .staticif(usedDepth, _.assign(curDepth, stackDepthPath))
+        .rest(newBody)
+    else
+      val resSym = TempSymbol(None, "stackDelayRes")
+      val rewritten = blockBuilder
+        .staticif(usedDepth, _.assign(curDepth, stackDepthPath))
+        .assignFieldN(runtimePath, STACK_DEPTH_IDENT, op("+", stackDepthPath, intLit(increment)))
         .assign(resSym, Call(checkDepthPath, Nil)(true, true))
+        .ifthen(
+          resSym.asPath,
+          Case.Cls(paths.effectSigSym, paths.effectSigPath),
+          Return(
+            // Call(Value.Lit(Tree.StrLit(doUnwindPath.get.toString())), resSym.asPath.asArg :: intLit(0).asArg :: Nil)(true, false),
+            Call(doUnwindPath.get, resSym.asPath.asArg :: intLit(0).asArg :: Nil)(true, false),
+            false
+          )
+        )
         .rest(newBody)
-  
-  def rewriteFn(defn: FunDefn) = FunDefn(defn.owner, defn.sym, defn.params, rewriteBlk(defn.body))
+      // float out defns to move the doUnwind function behind 
```

Comment:
Do you mean something like:
```suggestion
      // Float out defns, including the doUnwind function, so that they appear at the top of the block
```
?

Side note: It's a bit strange to do the floating out here. Maybe we should float things out later in the compilation pipeline.